File: cs/Chess/TemporaryPieceMove.cs
using System;

namespace Chess
{
    public class TemporaryPieceMove : IDisposable
    {
        private readonly Board board;
        private readonly Location from;
        private readonly Piece oldDestinationPiece;
        private readonly Location to;

        public TemporaryPieceMove(Board board, Location from, Location to, Piece oldDestinationPiece)
        {
            this.board = board;
            this.from = from;
            this.to = to;
            this.oldDestinationPiece = oldDestinationPiece;
        }

        public void Undo()
        {
            board.SetPieceToLocation(from, board.GetPiece(to));
            board.SetPieceToLocation(to, oldDestinationPiece);
        }

        public void Dispose()
        {
            Undo();
        }
    }
}

File: java/src/main/java/ru/kontur/courses/chess/TemporaryPieceMove.java
package ru.kontur.courses.chess;

import java.io.Closeable;
import java.io.IOException;

public class TemporaryPieceMove implements Closeable {
    private final Board board;
    private final Location from;
    private final Piece oldDestinationPiece;
    private final Location to;

    public TemporaryPieceMove(Board board, Location from, Location to, Piece oldDestinationPiece) {
        this.board = board;
        this.from = from;
        this.to = to;
        this.oldDestinationPiece = oldDestinationPiece;
    }

    public void undo() {
        board.set()
    }

    @Override
    public void close() throws IOException {
        undo();
    }
}
